### PR TITLE
Improve saving

### PR DESCRIPTION
**Pull request by ondrowan**
This patch contains all changes needed before the exhibition. There's:

* "new project"
* "save as..."
* handling of errors while saving / opening
* prompt if user wants to exit / open / create new project while the current project is unsaved

One thing that's still not done is layout of buttons in "prevent exit" dialog (save, don't save, cancel). I checked VSCode and they use "don't save" on the left side, then a gap and then "cancel" and "save" next to each other. Is this desirable?

### CODE REVIEW

File: src/lib.rs
Comment (yanchith) [inline src/lib.rs]: Just note that this has been renamed to "H.U.R.B.A.N. selector" meanwhile in master (notice the lowercase "s"). Also everywhere in this PR.

File: src/ui.rs
Comment (yanchith) [inline src/ui.rs]: Let's do `let changed = change.is_some()` prior to this and then return that from the function, so that we don't have to clone.

File: src/ui.rs
Comment (yanchith) [inline src/ui.rs]: Can we still try to match on a reference here and not clone?

File: src/project.rs
Comment (yanchith) [inline src/project.rs:40]: Let's add `impl error::Error for ProjectError {}`

File: src/lib.rs
Comment (yanchith) [inline src/lib.rs:426]: This probably shouldn't be here. The recomputing should happen when new results are produced by the interpreter. It will either happen automatically (when autorun is enabled), or manually. Either way, we probably shouldn't do it here, before the geometries in the scene actually change.

This is also true for new project. (I am sorry mislead you here in your previous PR)

File: src/lib.rs
Comment (yanchith) [inline src/lib.rs]: Shadowing is fine. Also one is `override` and one is `overwrite`. Maybe this should be unified across the PR.

File: src/project.rs
Comment (yanchith) [inline src/project.rs:16]: Let's implement `Copy` here. Clippy should afterwards point out places where this is cloned (there's a few cases in lib.rs)

**Review by yanchith — COMMENTED** [review]
Functionality seems to be working well :tada: 

Note that there's conflicts with master, they shouldn't be hard to resolve, but I will have to review again. @janper should probably review only once conflicts have been resolved. Also the code seems to be a bit more complex than what one would think it needs to be, but we can probably refactor later, since an obvious simplification didn't come to mind.

Nit: we'll have to remove the uses of `format!` macro from the UI eventually (per frame allocations are not very perf friendly and fragment the heap if the allocator isn't very good at de-fragmenting). All the strings there could be defined statically. Can wait for later though (I'll create a maintenance task in trello)

**Review by janper — CHANGES_REQUESTED** [review]
Consider approved, however these are my suggestions:
- **Unsaved changes** popup could put the three buttons next to each other?
- **Unsaved changes** popup should implement changes I made to Screenshot popup in PR#146 if that one gets merged first
- **Unsaved changes** popup - please rename the button from _Don't save_ to _Discard changes_
- **Unsaved changes** popup - please change button order to _Save_, _Discard changes_, _Cancel_
- Drop the word _project_ from all the buttons
-  (May not belong to this PR, but where else?) Please remove _Screenshot_ button to _Save Screenshot_
Thanks!

File: src/lib.rs
Comment (yanchith) [reply src/lib.rs:426]: Hmm, maybe we actually do want to keep this. Looks like a UX decision, so perhaps @janper should decide that. Do we want to clear the scene immediately when opening/newing a project? Or do we want to wait for the interpreter to run the new (potentially empty) program?

Comment (ondrowan) [reply src/lib.rs:426]: I have kept this as is until you two decide what to do with it.

**Comment by ondrowan** [comment]
This is reviewable again. Merge with `master` wasn't very smooth, so please check if I haven't accidentally removed something.

@janper There are some missing tooltips and notifications now that I've merged with `master` branch. Let's wait until everything else is approved and then we can add those as the last thing. Either tell me what texts to use or commit it yourself afterwards.

@yanchith I've removed the most offending `format!` that allocated `String` each frame and used `[].join()` instead. AFAIK it still allocates `String` but with pre-computed length, making it slightly faster. I guess you had something else in mind, so just let me know in chat and we can refactor it later. Since I had no idea what are the possibilities, I made some research and found [this juicy thread](https://www.reddit.com/r/rust/comments/48fmta/seven_ways_to_concatenate_strings_in_rust_the/).

**Review by janper — APPROVED** [review]

**Comment by janper** [comment]
Good enough. I'll need to review all the texts again (the printed description uses different words), so I'll take care of the hints as well.

One little thing, if possible - can the buttons be aligned to the window center? 
image
I can also do that in my upcoming PR.

File: src/lib.rs
Comment (yanchith) [reply src/lib.rs:426]: @janper Didn't reply. Let's keep it.

Comment (janper) [reply src/lib.rs:426]: Sorry, I didn't get any notification.
Also I don't know what does this code do.

I believe when a project is loaded, the scene should be wiped and set to default state. The geometry should be generated only when the pipeline is executed (either by a button or automatically).

**Comment by yanchith** [comment]
@janper 

> I can also do that in my upcoming PR.

Please do.

@ondrowan 

> I've removed the most offending format! that allocated String each frame and used [].join() instead.

That one didn't allocate per frame. The `change_window_title` function was only called conditionally. There are many other invcations of `format!` macro in UI that do it uncoditionally. Let's address that in other PR though.

If there are no objections, I am merging this soon.

**Review by yanchith — APPROVED** [review]

File: src/lib.rs
Comment (yanchith) [reply src/lib.rs:426]: Alright. I'll create a task to ensure this

**Pull request merged**
